contracts/VolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,152 @@
+    // check the imbalance in the sliding window
+    require(startBlock <= endBlock);
+
+    buyImbalance = 0;
+
+    for(uint windowInd = 0 ; windowInd < SLIDING_WINDOW_SIZE ; windowInd++ ) {
+      TokenImbalanceData memory perBlockData = tokenImbalanceData[token][windowInd];
+      if( perBlockData.lastBlock <= endBlock && perBlockData.lastBlock >= startBlock ) {
+        buyImbalance += int(perBlockData.lastBlockBuyUnitsImbalance);
+      }
+    }
+  }
+
+  function getImbalanceSincePriceUpdate( ERC20 token, uint priceUpdateBlock, uint currentBlock )
+    public view returns(int buyImbalance, int currentBlockImbalance) {
+    buyImbalance = 0;
+    currentBlockImbalance = 0;
+
+    uint64 latestBlock = uint64(0);
+
+    for(uint windowInd = 0 ; windowInd < SLIDING_WINDOW_SIZE ; windowInd++ ) {
+      TokenImbalanceData memory perBlockData = tokenImbalanceData[token][windowInd];
+      if( uint(perBlockData.lastPriceUpdateBlock) != priceUpdateBlock ) continue;
+      if( perBlockData.lastBlock < latestBlock ) continue;
+
+      latestBlock = perBlockData.lastBlock;
+      buyImbalance = perBlockData.totalBuyUnitsImbalance;
+      if( uint(perBlockData.lastBlock) == currentBlock ) {
+        currentBlockImbalance = perBlockData.lastBlockBuyUnitsImbalance;
+      }
+    }
+
+    if( buyImbalance == 0 ) {
+      buyImbalance = getImbalanceInRange( token, priceUpdateBlock, currentBlock );
+    }
+  }
+
+  function getImbalance( ERC20 token, uint priceUpdateBlock, uint currentBlock)
+    public view returns(int totalImbalance, int currentBlockImbalance) {
+
+    int resolution = int(tokenControlInfo[token].minimalRecordResolution);
+
+    (totalImbalance,currentBlockImbalance) = getImbalanceSincePriceUpdate(token,
+                                                                          priceUpdateBlock,
+                                                                          currentBlock );
+    totalImbalance *= resolution;
+    currentBlockImbalance *= resolution;
+  }
+
+
+  function addImbalance( ERC20 token,
+                         int buyAmount,
+                         uint priceUpdateBlock,
+                         uint currentBlock ) internal {
+
+    uint currentBlockIndex = currentBlock % SLIDING_WINDOW_SIZE;
+    int64 recordedBuyAmount = int64(buyAmount /
+                         int(tokenControlInfo[token].minimalRecordResolution));
+
+    int prevImbalance = 0;
+
+    TokenImbalanceData memory currentBlockData = tokenImbalanceData[token][currentBlockIndex];
+    // first scenario - this is not the first tx in the current block
+    if( currentBlockData.lastBlock == currentBlock ) {
+      if( uint(currentBlockData.lastPriceUpdateBlock) == priceUpdateBlock ){
+        // just increase imbalance
+        currentBlockData.lastBlockBuyUnitsImbalance += recordedBuyAmount;
+        currentBlockData.totalBuyUnitsImbalance += recordedBuyAmount;
+      }
+      else {
+        // imbalance was changed in the middle of the block
```

Comment:
I think remark should be - price was changed...

Reply:
actually both were changed in the middle. you are right.